packages/flutter/lib/src/widgets/router.dart

```diff
@@ -0,0 +1,1257 @@
+  ///    information even if location does change.
+  static void navigate(BuildContext context, VoidCallback callback) {
+    final _RouterScope scope = context
+      .getElementForInheritedWidgetOfExactType<_RouterScope>()
+      .widget as _RouterScope;
+    scope.routerState._setStateWithExplicitReportStatus(_IntentionToReportRouteInformation.must, callback);
+  }
+
+  /// Forces the [Router] to to run the [callback] without reporting the route
+  /// information back to the engine.
+  ///
+  /// Use this method if you don't want the [Router] to report the new route
+  /// information even if it detects changes as a result of running the
+  /// [callback].
+  ///
+  /// The web application relies on the [Router] to report new route information
+  /// in order to create browser history entry. The [Router] will report them
+  /// automatically if it detects the [RouteInformation.location] changes. You
+  /// can use this method if you want to navigate to a new route without
+  /// creating the browser history entry.
+  ///
+  /// See also:
+  ///
+  ///  * [Router]: see the "URL updates for web applications" section for more
+  ///    information about route information reporting.
+  ///  * [navigate]: which forces the [Router] to report the route information
+  ///    even if location does not change.
+  static void neglect(BuildContext context, VoidCallback callback) {
+    final _RouterScope scope = context
+      .getElementForInheritedWidgetOfExactType<_RouterScope>()
+      .widget as _RouterScope;
+    scope.routerState._setStateWithExplicitReportStatus(_IntentionToReportRouteInformation.ignore, callback);
+  }
+
+  @override
+  State<Router<T>> createState() => _RouterState<T>();
+}
+
+typedef _AsyncPassthrough<Q> = Future<Q> Function(Q);
+
+// Whether to report the route information in this build cycle.
+enum _IntentionToReportRouteInformation {
+  // We haven't receive any signal on whether to report.
+  none,
+  // Report if route information changes.
+  maybe,
+  // Report regardless of route information changes.
+  must,
+  // Don't report regardless of route information changes.
+  ignore,
+}
+
+class _RouterState<T> extends State<Router<T>> {
+  Object _currentRouteInformationParserTransaction;
+  Object _currentRouterDelegateTransaction;
+  _IntentionToReportRouteInformation _currentIntentionToReporting;
+
+  @override
+  void initState() {
+    super.initState();
+    widget.routeInformationProvider?.addListener(_handleRouteInformationProviderNotification);
+    widget.backButtonDispatcher?.addCallback(_handleBackButtonDispatcherNotification);
+    widget.routerDelegate.addListener(_handleRouterDelegateNotification);
+    _currentIntentionToReporting = _IntentionToReportRouteInformation.none;
+    if (widget.routeInformationProvider != null) {
+      _processInitialRoute();
+    }
+    _lastSeenLocation = widget.routeInformationProvider?.value?.location;
+  }
+  Object _scheduledRouteInformationReportTask;
```

Comment:
nit: separate some of these by blank lines? (at least between the end of initState and this line)